examples/nft/assets/nft_approval.py

```diff
@@ -0,0 +1,99 @@
+from pyteal import *
+
+def approval_program():
+    """
+    This smart contract implements a Non Fungible Token 
+    Each NFT is an (ID, ref_data, hash) triple.
+    NFT ID is the total count of NFT's in asc.
+    ref-data is represented as "https://nft-name/<total>/nft-ref"
+    hash is an hash of the external resources data.
+
+    Commands:
+        create    Creates a new NFT. Expects two additional arguments - nft-name, nft-ref 
+        tranfer   Transfers an NFT between two accounts. Expects one additional arg - NFT_ID.
+                  Additionally, two Accounts(from, to) must also be passed to the smart contract. 
+    """
+
+    # Check to see that the application ID is not set, indicating this is a creation call.
+    # Store the creator address to global state.
+    # Set total nft count to 0
+    on_creation = Seq([
+        App.globalPut(Bytes("creator"), Txn.sender()),
+        Assert(Txn.application_args.length() == Int(0)),
+        App.globalPut(Bytes("total"), Int(0)),
+        Return(Int(1))
+    ])                  
+
+	# Checks whether the sender is creator.
+    is_creator = Txn.sender() == App.globalGet(Bytes("creator"))
+
+    # Get total count of NFT's from global storage
+    total_nft = Itob(App.globalGet(Bytes("total")))
+
+    #create ref_data for NFT (https://nft-name/<total>/nft-ref)
+    ref_data = Concat(  
+        Bytes("https://"), 
+        Txn.application_args[1], # nft-name
+        Bytes(".com/"), 
+        Itob(App.globalGet(Bytes("total"))), 
+        Bytes("/"), 
+        Txn.application_args[2] # nft-ref
+    )
+
+    # Verifies if the creater is making this request
+    # Verifies three arguments are passed to this transaction ("create", nft-name, nft-ref)
+    # Increment Global NFT count by 1
+    # Assign ref_data to NFT ID (= total)
+    # Assign hash of nft-ref to ID_h
+    # Add above two keys to global and creator's local storage
+    create_nft = Seq([
+        Assert(is_creator),
+        Assert(Txn.application_args.length() == Int(3)),
+
+        App.globalPut(Bytes("total"), App.globalGet(Bytes("total")) + Int(1)),
+
+        App.globalPut(total_nft, ref_data),
+        App.globalPut(Concat(total_nft, Bytes("_h")) , Sha256(Txn.application_args[2])),
+
+        App.localPut(Int(0), total_nft, ref_data),  # Int(0) represents the address of caller
+        App.localPut(Int(0), Concat(total_nft, Bytes("_h")) , Sha256(Txn.application_args[2])),
```

Comment:
1. Let's store `Concat(total_nft, Bytes("_h"))` in a scratch and reuse it
2. Same for `Sha256(Txn.application_args[2])` - this operation is expensive!

Reply:
I have cleaned up the construction a little bit. I saw that pyTEAL provides functions for scratch space (`ScratchLoad`, `ScratchStore` and `ScratchSlot`). But i coudn't find any reference or examples to use it. I am looking into it. 
[](https://pyteal.readthedocs.io/en/stable/api.html#pyteal.ScratchSlot)

Reply:
Will update after: [https://github.com/algorand/pyteal/pull/33/files#diff-71c21b6210f42dca46ebf000fbc3425e305bea125790e986b2b6debe4c8473f8R17]()